### PR TITLE
chore(site): add remote playwright support and script

### PR DESCRIPTION
**Pull request by mafredri**
Got inspired to play around with this, and since I tried most of this out already, thought I'd make a PR.

I have no preference on the final implementation, so feel free to change this as you please.

Adds a new command `pnpm run playwright:remote` that starts a playwright server _locally_ and opens up port forwards to the provided workspace and drops you into an SSH shell where you can start the test.

(This relies on `config-ssh` to use plain OpenSSH to simplify things.)

image

### CODE REVIEW

**Review by BrunoQuaresma — COMMENTED** [review]
Even after reading the changes Idk how to use this, maybe you could add some docs about this in the docs/contributing/fe?

**Review by BrunoQuaresma — APPROVED** [review]
I thought it was going to allow me to run the test on my workspace using the browser on my local machine but if this is an improvement, I'm up for it.

**Comment by mafredri** [comment]
> I thought it was going to allow me to run the test on my workspace using the browser on my local machine

That's exactly what this does.

File: site/e2e/server/server.mjs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+import { chromium } from "@playwright/test";
+
+const server = await chromium.launchServer({ headless: false });
```
Comment (mtojek) [inline site/e2e/server/server.mjs:3]: I suspect that .mjs extension is on purpose?

File: site/e2e/server/run_with_forward.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+workspace=${1:-}
+port=${2:-3000}
+
+if [[ -z "${workspace}" ]]; then
+  echo "Usage: $0 <workspace> [port]"
+  exit 1
+fi
+
+# Go to site.
+cd "$(dirname "$0")/../.."
```
Comment (mtojek) [inline site/e2e/server/run_with_forward.sh:13]: if the script breaks, will the user land in another dir?

File: site/e2e/server/run_with_forward.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+read -r ws_endpoint <&3
+if [[ ${ws_endpoint} != ws://* ]]; then
+  echo "Playwright failed to start."
+  echo "${ws_endpoint}"
+  cat "$playwright_out"
+  exit 1
+fi
+echo "Playwright started at ${ws_endpoint}"
+
+ws_port=${ws_endpoint##*:}
```
Comment (mtojek) [inline site/e2e/server/run_with_forward.sh:47]: I'm curious, is something like this supported: `ws_port=${ws_endpoint##*:*/}`?

File: site/e2e/server/server.mjs (hunk 1 above)
Comment (mafredri) [reply site/e2e/server/server.mjs:3]: When running in Node, this was the only way I could make `import`-statements work (vs `require`), I also tried `tsnode` with `.ts` but that was a bust. Open for alt. suggestions.

File: site/e2e/server/run_with_forward.sh (hunk 2 above)
Comment (mafredri) [reply site/e2e/server/run_with_forward.sh:13]: Nah, only if someone tries to source this script which is not an intended use-case.

File: site/e2e/server/run_with_forward.sh (hunk 3 above)
Comment (mafredri) [reply site/e2e/server/run_with_forward.sh:47]: Not that I know of, would be neat though.

File: site/e2e/server/run_with_forward.sh (hunk 2 above)
Comment (mtojek) [reply site/e2e/server/run_with_forward.sh:13]: It would be cool to add a script-check to verify if it isn't directly called, but definitely not a blocker

**Review by mtojek — APPROVED** [review]

File: site/e2e/server/run_with_forward.sh (hunk 2 above)
Comment (mafredri) [reply site/e2e/server/run_with_forward.sh:13]: Tbh I think I should just rewrite this completely to remove the coder repo requirement locally. Feels like it’d just add confusion trying to document this. Essentially you could either run the script from the repo or download it and run it individually (curl | bash), we can then just pick the playwright version requirement directly from the workspace which mean it has a greater chance of working on custom branches too.

**Comment by BrunoQuaresma** [comment]
> > I thought it was going to allow me to run the test on my workspace using the browser on my local machine
> 
> That's exactly what this does.

So why do I need to install extra things - from what I could see - besides Chrome only?

**Comment by mafredri** [comment]
> So why do I need to install extra things - from what I could see - besides Chrome only?

Because playwright decided to write their own API vs using plain Chrome DevTools Protocol. 😔

**Comment by mafredri** [comment]
@BrunoQuaresma @mtojek I rewrote the script/setup a bit now, added docs, hopefully it's a bit clearer and easier to use.

**Review by mtojek — APPROVED** [review]
👍 You may want to record a demo and post it on slack :)

**Pull request merged**